chainstate/src/detail/chainstateref/consistency_checker.rs

```diff
@@ -0,0 +1,238 @@
+            chain_config,
+            block_map_keys,
+            block_index_map,
+            block_by_height_map,
+            best_block_id,
+            min_height_with_allowed_reorg,
+        })
+    }
+
+    pub fn check(&self) {
+        log::debug!("Running chainstate consistency checks");
+
+        self.check_block_index_consistency();
+        self.check_block_height_map_consistency();
+
+        // TODO: add consistency checks for other maps in the chainstate db.
+        // https://github.com/mintlayer/mintlayer-core/issues/1710
+    }
+
+    /// Check the block map vs block index map consistency.
+    fn check_block_index_consistency(&self) {
+        // Loop over block_map_keys and block_index_map simultaneously via merge_join_by, looking
+        // for ids that are present in one of them and missing in the other.
+        for merged in self
+            .block_map_keys
+            .iter()
+            .merge_join_by(self.block_index_map.iter(), |id1, (id2, _)| {
+                Ord::cmp(id1, id2)
+            })
+        {
+            let (block_id, block_index) = match merged {
+                EitherOrBoth::Left(block_id) => {
+                    // The block object is present, the index object is not.
+                    panic!("{PANIC_MSG}: block index data missing for block {block_id}");
+                }
+                EitherOrBoth::Right((block_id, block_index)) => {
+                    // The block index object is present, the block object is not;
+                    // The persistence flag must be unset and the status must not be "ok".
+                    assert!(
+                        !block_index.is_persisted(),
+                        "{PANIC_MSG}: block {block_id} must not be persisted"
+                    );
+                    assert!(
+                        !block_index.status().is_ok(),
+                        "{PANIC_MSG}: block {block_id} must not be ok"
+                    );
+
+                    (block_id, block_index)
+                }
+                EitherOrBoth::Both(_, (block_id, block_index)) => {
+                    // Both the block and block index objects are present.
+
+                    // The persistence flag must be set.
+                    assert!(
+                        block_index.is_persisted(),
+                        "{PANIC_MSG}: block {block_id} must be persisted"
+                    );
+
+                    (block_id, block_index)
+                }
+            };
+
+            let block_id_in_block_index = block_index.block_id();
+            assert_eq!(
+                block_id,
+                block_id_in_block_index,
+                "{PANIC_MSG}: block id from BlockIndex {block_id_in_block_index} doesn't match {block_id}"
+            );
+
+            // Check the parent, if it's not genesis.
+            if let Some(parent_id) =
+                block_index.prev_block_id().classify(self.chain_config).chain_block_id()
```

Comment:
Also check if `prev_block_id` is `None`, the current gen-block is genesis?

Reply:
Maybe also check parent block id in the index matches parent block id in the block.

Reply:
> Also check if prev_block_id is None, the current gen-block is genesis?

`prev_block_id` can't be `None`, because we're dealing with the "non-gen" `BlockIndex` here. Or do you mean to check the case when `chain_block_id` returns `None` here? In that case IMO it'll be a redundant check (because `classify` returns `None` only when the block id equals the genesis id obtained from chain config).

Reply:
> Maybe also check parent block id in the index matches parent block id in the block.

I decided to compare the whole header then.
But note that either way this check requires loading the whole block from the db; this makes some tests run noticeably longer in debug builds (like 60s becomes 160s). So I enable this check only in release builds.